realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/SessionTests.kt

```diff
@@ -0,0 +1,562 @@
+                    resources.close()
+                    handler.executeClientReset()
+
+                    // Validate that files have been moved
+                    Assert.assertFalse(handler.originalFile.exists())
+                    Assert.assertTrue(handler.backupFile.exists())
+                    looperThread.testComplete()
+                }
+                .build()
+        val realm = Realm.getInstance(config)
+        resources.add(realm)
+
+        // Trigger error
+        user.app.sync.simulateClientReset(app.sync.getSession(configuration))
+    }
+
+    // Check that we can use the backup SyncConfiguration to open the Realm.
+    @Test
+    fun errorHandler_useBackupSyncConfigurationForClientReset() = looperThread.runBlocking {
+        val resources = ResourceContainer()
+        val config = configFactory.createSyncConfigurationBuilder(user)
+                .clientResyncMode(ClientResyncMode.MANUAL)
+                .schema(StringOnly::class.java)
+                .errorHandler { session: SyncSession?, error: ObjectServerError ->
+                    if (error.errorCode != ErrorCode.CLIENT_RESET) {
+                        Assert.fail("Wrong error $error")
+                        return@errorHandler
+                    }
+                    val handler = error as ClientResetRequiredError
+                    // Execute Client Reset
+                    resources.close()
+                    handler.executeClientReset()
+
+                    // Validate that files have been moved
+                    Assert.assertFalse(handler.originalFile.exists())
+                    Assert.assertTrue(handler.backupFile.exists())
+                    val backupRealmConfiguration = handler.backupRealmConfiguration
+                    Assert.assertNotNull(backupRealmConfiguration)
+                    Assert.assertFalse(backupRealmConfiguration.isSyncConfiguration)
+                    Assert.assertTrue(backupRealmConfiguration.isRecoveryConfiguration)
+                    val backupRealm = Realm.getInstance(backupRealmConfiguration)
+                    Assert.assertFalse(backupRealm.isEmpty)
+                    Assert.assertEquals(1, backupRealm.where(StringOnly::class.java).count())
+                    backupRealm.where(StringOnly::class.java).findAll().first()?.let { first ->
+                        Assert.assertEquals("Foo", first.chars)
+                    } ?: fail()
+                    backupRealm.close()
+
+                    // opening a Dynamic Realm should also work
+                    val dynamicRealm = DynamicRealm.getInstance(backupRealmConfiguration)
+                    dynamicRealm.schema.checkHasTable(StringOnly.CLASS_NAME, "Dynamic Realm should contains " + StringOnly.CLASS_NAME)
+                    val all = dynamicRealm.where(StringOnly.CLASS_NAME).findAll()
+                    Assert.assertEquals(1, all.size.toLong())
+                    all.first()?.let { first ->
```

Comment:
IMO tests should use `!!` instead of `?` as we are otherwise hiding errors

Reply:
👍  I was bit too eager in avoid it. I already fixed it locally as the assert could then also be a one liner.